Conversion to Gradle KTS

Doomsdayrs opened the pull request:
<!-- Please describe your changes on the following line: -->

Convert Android Groovy Script to Kotlin Script.

As to possibly facilitate #33742 and or to simply match current Android Standards.

These are history destructive commits, as git will not recognize such heavy modifications as file moves.

All Logic is preserved per each commit sequentially. This means the first commit to the last preserves logic as to allow the first commit to be merged without the following ones, but not as to allow the following commits to be merged without the prior commit. 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build --debug --target x86_64-linux-android` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Doomsdayrs committed:
Convert settings.gradle to Kotlin Script

Stage 1 of #33742

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
servoview-local: Convert build.gradle to Kotlin Script

Stage 1 of #33742

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
Convert build.gradle to Kotlin Script

This was a trickier one, as I wanted to maintain compatibility with the rest of the files while facilitating this migration.

Closures are annoying, another annoyance of loosely typed languages in an OOP project.

Migration of child build scripts will require the reverse code and or migration of this scripts functions to kotlin lambdas / functions (which are just jvm functions).

Code based off of the following guide.
https://docs.gradle.org/current/userguide/kotlin_dsl.html#groovy_closures_from_kotlin

Stage 1 of #33742

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
servoapp: Convert build.gradle to Kotlin Script

Migrated deprecated API usages.
There are two more, but ignored for now.
("splits.density", "capitalize")

Stage 1 of #33742

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
servoview: Convert build.gradle to Kotlin Script

Migrated deprecated API usages.
There are two more, but ignored for now.
("splits.density", "capitalize")

Stage 1 of #33742

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs commented on support/android/apk/servoview/build.gradle.kts:
```diff
@@ -0,0 +1,259 @@
+
+    var result = ""
+
+    ResourceGroovyMethods.eachFileRecurse(path, FileType.FILES, closureOf<File> {
+        if (this.name == filename) {
+            result = this.absolutePath
+        }
+    })
```
This does not work, trying to find a replacement.

Doomsdayrs replied:
Maybe it did work, the replacement is also failing to work though..

Doomsdayrs replied:
Figured it out, `if (path.isNotBlank())` was missing.

Doomsdayrs committed:
servoview: Replace ResourceGroovyMethods with Kotlin File.walk

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs commented:
Alright, with that all the build code is now in Kotlin Script.

Next steps, Kotlinifying the various functions to no longer require Groovy Closures.

Doomsdayrs committed:
Replace Groovy Closures with Kotlin Lambda types

Stage 1 of #33742

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
Move Utility fields to buildSrc

Using extra fields is quite annoying and makes it hard to maintain
 API stability.

"buildSrc" is designed for this task, and thus is being used
 for said task.

This means that when editing build.gradle files in an Android Studio,
 there is a direct reference to the source of a function.
 (Easier time referring to documentation, source of function, etc).

More information here:
https://docs.gradle.org/current/userguide/sharing_build_logic_between_subprojects.html

Stage 1 of #33742

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs commented:
I personally feel that I cannot improve this any further currently.

Doomsdayrs commented:
Just found out that github butchered verifications when rebasing.

When I get home I'll repush to fix the signing.

Doomsdayrs commented:
May someone enable the android CI for this PR?

mrobinson commented:
```
Daemon will be stopped at the end of the build 
> Task :buildSrc:checkKotlinGradlePluginConfigurationErrors
> Task :buildSrc:pluginDescriptors
> Task :buildSrc:processResources NO-SOURCE
> Task :buildSrc:compileKotlin
> Task :buildSrc:compileJava NO-SOURCE
> Task :buildSrc:compileGroovy NO-SOURCE
> Task :buildSrc:classes UP-TO-DATE

> Task :buildSrc:jar
:jar: No valid plugin descriptors were found in META-INF/gradle-plugins

> Configure project :servoapp
w: file:///home/runner/work/servo/servo/support/android/apk/servoapp/build.gradle.kts:153:69: 'capitalize(): String' is deprecated. Use replaceFirstChar instead.

WARNING: API 'splits.density' is obsolete.
FAILURE: Build failed with an exception.

* Where:
Build file '/home/runner/work/servo/servo/support/android/apk/servoapp/build.gradle.kts' line: 55

* What went wrong:
SigningConfig with name 'release' not found.
```

Doomsdayrs commented:
Ah, I know the solution for that.

(using "named" instead of "register" in build script)

Doomsdayrs commented:
May someone try the android run again?

Doomsdayrs commented:
Ah, Google play requires API 33 or higher. A tomorrow task, for now I sleep.

Doomsdayrs committed:
Sync target SDK to 33

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs commented:
Request for a test again.

Doomsdayrs commented:
Touched the sun with that.

Doomsdayrs commented:
Hopefully the final attempt.

Doomsdayrs commented:
😮‍💨 Alright, so it wasn't because of API 34 being the target.

the solution is to correct the notification code, which I don't want to do due to maintaining scope of this PR, but seems I have to.

Doomsdayrs committed:
Make Notification actions immutable.

Otherwise android lint will be upset.

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
Move dependencies from servoview to servoapp

ServoView does not use them.

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
Add POST_NOTIFICATIONS to manifest

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs committed:
Add host to intent-filter

Use "*" for any host, lets hope this works.

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs commented:
I went out of scope and fixed lint issues as well. Hopefully now an android build will work! (it did on my PC before too, but :shrug: )

Doomsdayrs commented:
PRAISE BE!

mukilan commented:
Great work!

I was testing this change locally and it crashes on my device. It seems like the APK produced doesn't contain the native shared objects (libservoshell.so and libc++_shared.so). The [artifacts generated](https://github.com/servo/servo/actions/runs/11412144557) from the recent try run also confirm this - the APK is too small and doesn't have the 'lib' folder inside it.

I can investigate the issue later this week, but if you wanted to look into it, I suspect [this logic to copy the native objects](https://github.com/servo/servo/blob/4d11b2dc84d53b90ca5f2b729ffa1a5db2da45e9/support/android/apk/servoview/build.gradle#L144) is not being triggered as expected, perhaps due to a change in the name of the task we are looking to match using the regex.

Doomsdayrs commented:
> Great work!
> 
> I was testing this change locally and it crashes on my device. It seems like the APK produced doesn't contain the native shared objects (libservoshell.so and libc++_shared.so). The [artifacts generated](https://github.com/servo/servo/actions/runs/11412144557) from the recent try run also confirm this - the APK is too small and doesn't have the 'lib' folder inside it.
> 
> I can investigate the issue later this week, but if you wanted to look into it, I suspect [this logic to copy the native objects](https://github.com/servo/servo/blob/4d11b2dc84d53b90ca5f2b729ffa1a5db2da45e9/support/android/apk/servoview/build.gradle#L144) is not being triggered as expected, perhaps due to a change in the name of the task we are looking to match using the regex.

Something something `tasks.forEach` is occurring before the merge task is completed. Something something now have to figure out where it goes. I am investigating this right now.

Doomsdayrs committed:
Solve ndkBuild tasks not being linked

The problem stems from something something groovy wishy washy unclear
 execution order something Kotlin explicit execution order.

Merge tasks exist after the project is evaluated.
 The problem is that simply running afterEvaluate causes an
 ConcurrentModificationException. This is because of creating a new
 task while looping over existing tasks. To remedy this we simply
 filter the tasks first, than create and link the new task.

Signed-off-by: clocks <doomsdayrs@gmail.com>

Doomsdayrs commented:
Maestro, another build?

Doomsdayrs commented on support/android/apk/buildSrc/src/main/kotlin/Interop.kt:
```diff
@@ -0,0 +1,67 @@
+    return basePath + "/target/android/" + getSubTargetDir(debug, arch)
+}
+
+fun Project.getNativeTargetDir(debug: Boolean, arch: String) {
```
This does not return a string, thus the path ends up with a `Kotlin.Unit`

Doomsdayrs commented:
Perchance let the CI spin again?

Doomsdayrs commented:
It succeeded and also had a chunky artifact size, praise!

mukilan commented:
Thanks! I'll review this by the end of the week.

mukilan commented on support/android/apk/buildSrc/src/main/kotlin/Interop.kt:
```diff
@@ -0,0 +1,67 @@
+    return getRustTarget(arch) + "/" + if (debug) "debug" else "release"
+}
+
+fun Project.getJniLibsPath(debug: Boolean, arch: String): String =
```
Why are some functions defined on the `Project` but not others?

mukilan commented on support/android/apk/buildSrc/src/main/kotlin/Interop.kt:
```diff
@@ -0,0 +1,67 @@
+    return getRustTarget(arch) + "/" + if (debug) "debug" else "release"
+}
+
+fun Project.getJniLibsPath(debug: Boolean, arch: String): String =
```
Why are some functions defined on the Project but not others?

Doomsdayrs replied:
Unlike in groovy in which scope is implicit and not clear, Kotlin requires explicit scoping.

`Project.` means this function is an extension to `Project` classes, thus has access to public fields of the `Project` class, and can only be called from a project object or scope.

I shall add documentation to the top of the file in this regard.

Doomsdayrs replied:
[786edc0](https://github.com/servo/servo/pull/33772/commits/786edc07ad24676396fc98815c44e7dc70d8c879)

mukilan commented on support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt:
```diff
@@ -0,0 +1,20 @@
+
+val versionCodeString: String = SimpleDateFormat("yyyyMMdd00").format(today)
+
+val generatedVersionCode = versionCodeString.toInt()
```
Does Gradle guarantee that the logic in `buildSrc` will be run only once for both servoapp and servoview. My concern is that `servoapp` and `servoview` artifacts produced by a single build might not have the same version if that is not garaunteed.

Doomsdayrs replied:
From what I understand from experience and [documentation](https://docs.gradle.org/current/userguide/sharing_build_logic_between_subprojects.html), and [member](https://discuss.gradle.org/t/buildsrc-vs-build-logic/46708/2) of the gradle team, Yes it is only produced once.

mukilan replied:
Thanks for the links! It does look like Gradle guarantees that the `buildSrc` is **built once** but I am not sure if that also implies if the built code **executes** once, globally across all sub-projects. In any case, we should be safe for now as the version code only has a granularity of a day, but this might be something we need to keep in mind if we ever extend it to include minutes, seconds or some other transient state.

Doomsdayrs replied:
because the version code is assigned to a val, we can assume it compiles into a constant int and said class is just imported by other projects.

I'll resolve this for now.

mukilan commented on support/android/apk/servoapp/src/main/AndroidManifest.xml:
```diff
@@ -26,6 +27,7 @@
               <category android:name="android.intent.category.DEFAULT" />
               <category android:name="android.intent.category.BROWSABLE" />
               <data android:scheme="http" />
+              <data android:host="*"/>
```
What is the motivation for this change? Was it not working previously?

Doomsdayrs replied:
It was a part of a series of multiple commits addressing android lint errors, which in turn caused the CI to fail.

You may experiment on a local machine by dropping the commits and running the android lint task from gradle.

Doomsdayrs replied:
Checking in, Have you done the experiment yet?

mukilan replied:
Sorry about the delay. I did try removing `host` locally and it didn't fail the build for me. Do you have links to the logs from CI where it failed?

mukilan replied:
I did a [test run on the CI ](https://github.com/mukilan/servo/actions/runs/11625426495) without the host directives and it succeeded.  I'll update this PR to remove those. We can always add it back if we find it is needed.

Doomsdayrs committed:
Add documentation to why some functions are extensions to Project

Signed-off-by: clocks <doomsdayrs@gmail.com>

mukilan committed:
android: drop the host directives from AndroidManifest.xml

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan commented:
Great work! Thank you for working through all the issue with the migration!
